File: lib/domain/entities/repository.dart
import 'package:uuid/uuid.dart';

/// 仓库实体
class Repository {
  final String id;
  final String name;
  final String icon;
  final String color;
  final bool isArchived;
  final bool isDeleted;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Repository({
    required this.id,
    required this.name,
    this.icon = 'folder',
    this.color = '#3B82F6',
    this.isArchived = false,
    this.isDeleted = false,
    required this.createdAt,
    required this.updatedAt,
  });

  /// 创建新仓库（无 ID）
  factory Repository.create({
    required String name,
    String icon = 'folder',
    String color = '#3B82F6',
  }) {
    final now = DateTime.now();
    return Repository(
      id: const Uuid().v4(),
      name: name,
      icon: icon,
      color: color,
      createdAt: now,
      updatedAt: now,
    );
  }

  /// copyWith 方法
  Repository copyWith({
    String? id,
    String? name,
    String? icon,
    String? color,
    bool? isArchived,
    bool? isDeleted,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Repository(
      id: id ?? this.id,
      name: name ?? this.name,
      icon: icon ?? this.icon,
      color: color ?? this.color,
      isArchived: isArchived ?? this.isArchived,
      isDeleted: isDeleted ?? this.isDeleted,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  /// 序列化为 JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'icon': icon,
      'color': color,
      'isArchived': isArchived,
      'isDeleted': isDeleted,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Repository &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;
}

File: web/src/domain/entities/repository.ts
import { v4 as uuidv4 } from 'uuid';

export interface RepositoryData {
  id: string;
  name: string;
  description: string | null;
  defaultBranchId: string | null;
  icon: string;
  color: string;
  isArchived: boolean;
  isDeleted: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export class Repository implements RepositoryData {
  id: string;
  name: string;
  description: string | null;
  defaultBranchId: string | null;
  icon: string;
  color: string;
  isArchived: boolean;
  isDeleted: boolean;
  createdAt: Date;
  updatedAt: Date;

  constructor(data: RepositoryData) {
    this.id = data.id;
    this.name = data.name;
    this.description = data.description;
    this.defaultBranchId = data.defaultBranchId;
    this.icon = data.icon;
    this.color = data.color;
    this.isArchived = data.isArchived;
    this.isDeleted = data.isDeleted;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
  }

  static create(
    name: string,
    icon = 'repository',
    color = '#3B82F6',
    description: string | null = null
  ): Repository {
    const now = new Date();
    return new Repository({
      id: uuidv4(),
      name,
      description,
      defaultBranchId: null,
      icon,
      color,
      isArchived: false,
      isDeleted: false,
      createdAt: now,
      updatedAt: now,
    });
  }

  copyWith(updates: Partial<RepositoryData>): Repository {
    return new Repository({
      ...this,
      ...updates,
      updatedAt: new Date(),
    });
  }

  toJson(): Record<string, unknown> {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      defaultBranchId: this.defaultBranchId,
      icon: this.icon,
      color: this.color,
      isArchived: this.isArchived,
      isDeleted: this.isDeleted,
      createdAt: this.createdAt.toISOString(),
      updatedAt: this.updatedAt.toISOString(),
    };
  }
}
